enhancements/kube-apiserver/enforce-labels.md

```diff
@@ -0,0 +1,138 @@
+- an overkill for our immediate needs, provides a much broader functionality than we need
+
+NOTE: We are only interested in the planned mutating feature of the project.
+
+#### Gatekeeper operator
+
+Gatekeeper would be installed using an operator. The operator would not only take care of the installation, but also perform some additional configuration to ensure gatekeeper does not interere with control-plane resources. 
+
+There is currently an effort to provide a gatekeeper operator [here|https://github.com/font/gatekeeper-operator]. It is planned that the operator will be accepted as part of the open-policy-agent repo.
+
+Gatekeeper inspects resources based on a validating webhook (we assume the mutation feature will similarly use a mutating webhook). A sample webhook definition can be found [here|https://github.com/open-policy-agent/gatekeeper/blob/1de87b6d3c2ed3609e69b789d722e28285873861/charts/gatekeeper/templates/gatekeeper-validating-webhook-configuration-validatingwebhookconfiguration.yaml]. The operator would modify the webhook namespace selector to exclude all control-plane components. The operator would keep track of the namespaces and only allow the gatekeeper webhooks access to non control-plane namespaces. The operator configuration could allow to configure the namespaces accessible by gatekeeper.
+
+#### Safeguaring control-plane components
+
+Gatekeeper uses a single admission webhook to for constraints enforcement. We assume the same solution will be used for the mutation feature. The control-plane resources can be excluded from gatekeeper interference by specifying an appropriate namespace selector on the webhook.
+Gatekeeper would be installed/configured by an operator, which would adjust the webhook namespace selector (based on some GatekeeperInstallConfig CRD). 
+
+The webhook namespaceselector allows to filter the included namespaces based on the namespace labels (details available: [here|https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/] [here|https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/]) .
+
+Also the default gatekeeper webhook configuration is using "failurePolicy​: ​Ignore", we assume this would be extended to the mutation feature (more on this: https://github.com/open-policy-agent/gatekeeper#admission-webhook-fail-open-status). Any resources which would slip into the system unmodified would later be corrected using the additional relabeling feature (implemented in the second phase).
+
+
+#### Performance
+
+Multi-pod deployments of Gatekeeper is already implemented. This will allow to run multiple pods handling the mutating webhook to allow for better performance, better latency and resilience.
+The relabeling feature would preferably be implemented to run on multiple pods as well (daemonset?). The relabeling feature will most probably be implemented not as a webhook but as a controller.
```

Comment:
> The relabeling feature will most probably be implemented not as a webhook but as a controller.

This confuses me. You talk about a mutating webhook above and now you say it will be a controller? Or are you saying that you will only label on create via webhook, and via controller for updates?

Reply:
There will be 2 features:
- a mutated webhook mutating all created resources - this would only updated resources on creation, with the policies active at the moment of creation (iIn case the webhook is down, some resources could get into the system unmodified)
- a drift correction mechnism - this would updated existing resources if they are detected to be different than what the policy enforces. This would be activated whenever the policy changes, or after gatekeeper is restarted.

Currently there is a gatekeeper design doc only for the first feature. The second feature is still under discussion, with earliest estimate for any design being next year. I updated the PR to reflect the uncertainty of the implementation details of the second implementation. 

I also added a new section: "Gatekeeper features required" to make this more clear.

Reply:
sounds better. Am less concerned if only create requests get additional latency.